### ISSUE TITLE
Support for glide typing?

### ISSUE THREAD

**Issue by mahc9kez** [dslul/openboard#3]
AOSP keyboard have settings for glide typing but does not included in OpenBoard.

I know it's because glide typing require `libjni_latinimegoogle.so` the proprietary library (it can be extracted from [OpenGApps](https://sourceforge.net/projects/opengapps/files/) Optional/swypelibs-lib-arm64.tar.lz).

But maybe consider add this feature to Play Store version by using product flavors? Since play store users probably already have GApps on their phone.

I'm also trying to build LatinIME, but can't get glide typing works for me (but it works on LatinIME in LineageOS). it's a bit strange. so you may make sure at least it works on your devices if are about to support this feature.

**Comment by dslul** [comment]
I thought about this, but I'm hesitant to leave this dependency, even on those phones that have Google apps installed. The ideal thing would be to reimplement the swype logic in the library.

**Comment by mahc9kez** [comment]
AnySoftKeyboard have swype support which is FOSS, but the last time I tried it, it's far from good as gboard or any of other proprietary swype IME. I think it's not worth the effect to implement this feature for now.

But if you can add swype support and "joystick delete", it would make the keyboard much more usable. Joystick delete is a feature included in gboard but not enabled by default in stable (enabled by default in testing which require GApps to automatically get the config). With joystick delete, it show you a stick when you press the "Delete" icon, you can select words if you immediately swipe left, lift up to delete. or long press to quickly delete.

**Comment by machiav3lli** [comment]
Maybe implementing the space swipe gestures(select & delete) could be a good start point. It is already implemented in SimpleKeyboard (https://github.com/rkkr/simple-keyboard).

**Comment by mahc9kez** [comment]
oh I didn't know the _delete swipe_ feature in Simple Keyboard before, it's exactly the same thing gboard have, _joystick delete_ is just a visual stick which is not necessary.

**Comment by turtle836** [comment]
Moving the cursor by swiping on space would also be very nice.

**Comment by dslul** [comment]
> Moving the cursor by swiping on space would also be very nice.

This is already implemented together with delete swipe, but the F-Droid build system takes several days to compile the new versions. The Play Store version is already up to date and has those features.

**Comment by turtle836** [comment]
> > Moving the cursor by swiping on space would also be very nice.
> 
> This is already implemented together with delete swipe, but the F-Droid build system takes several days to compile the new versions. The Play Store version is already up to date and has those features.

Nice, thank you!

**Comment by smigima** [comment]
Hi,
Thanks for great work, after trying (unsuccessfully) to extract AOSP Keyboard from Lineage for several hours, it's refreshing to just install it from f-droid.

But about that gliding - is there any workaround I could do by myself to enable glide? 
I was trying to:
-Add libjni_latinimegoogle.so to arm64 in-app folder
But it didn't made a difference
-Remove libjni_latinime.so, copy libjni_latinimegoogle.so and rename it to libjni_latinime.so
But it crashed an app
-Add libjni_latinimegoogle.so to system/lib and system/lib64
But it didn't made a difference

I'm not asking you to implement some options including that lib, but maybe you can just don't cut it from AOSP Keyboard?
So when someone paste that lib into app's folder it would work?

**Comment by dslul** [comment]
> Hi,
> Thanks for great work, after trying (unsuccessfully) to extract AOSP Keyboard from Lineage for several hours, it's refreshing to just install it from f-droid.
> 
> But about that gliding - is there any workaround I could do by myself to enable glide?
> I was trying to:
> -Add libjni_latinimegoogle.so to arm64 in-app folder
> But it didn't made a difference
> -Remove libjni_latinime.so, copy libjni_latinimegoogle.so and rename it to libjni_latinime.so
> But it crashed an app
> -Add libjni_latinimegoogle.so to system/lib and system/lib64
> But it didn't made a difference
> 
> I'm not asking you to implement some options including that lib, but maybe you can just don't cut it from AOSP Keyboard?
> So when someone paste that lib into app's folder it would work?

I actually disabled swipe from code, so even if you put that library it won't work. If time permits, I want to reimplement the swipe logic (no ETA though).

**Comment by v0idifier** [comment]
@dslul is there a way to re enable swipe? i can re-compile myself.

**Comment by dslul** [comment]
> @dslul is there a way to re enable swipe? i can re-compile myself.

Just change [here](https://github.com/dslul/openboard/blob/master/app/src/main/java/org/dslul/openboard/inputmethod/latin/define/JniLibName.java) the library's name to jni_latinimegoogle and then enable gestures by setting [sHaveGestureLib=true](https://github.com/dslul/openboard/blob/master/app/src/main/java/org/dslul/openboard/inputmethod/latin/utils/JniUtils.java)

**Comment by v0idifier** [comment]
should i change `JNI_LIB_NAME` or uncomment the `JNI_LIB_NAME2` line?

**Comment by v0idifier** [comment]
changing `JNI_LIB_NAME` results in a crash.

**Comment by dslul** [comment]
> changing `JNI_LIB_NAME` results in a crash.

Do you have that library in your system?

**Comment by mahc9kez** [comment]
Set `JNI_LIB_NAME` to `jni_latinimegoogle` result in crash here (can't open the keyboard). Normal typing works correctly when uncomment JNI_LIB_NAME2 line, but crash after glide typing each time (the curve was shown but does not input anything when swype).

I'm using LineageOS 17.0 (Android 10) unofficial, maybe it's ROM related, I have tried to put libjni_latinimegoogle.so into system, but the built-in AOSP keyboard always crash when swype.

@dslul or anyone could try to remove libjni_latinimegoogle.so and include the so file in the apk and check if it could work?

`app/src/main/jniLibs/arm64-v8a/libjni_latinimegoogle.so`

**Comment by v0idifier** [comment]
> > changing `JNI_LIB_NAME` results in a crash.
> 
> Do you have that library in your system?

of course. i think it said something related to emoji in the logcat; maybe a mismatch in method names? i tried with the library from opengapps for android 10 and 9.

**Comment by v0idifier** [comment]
i've since read the source code of a different AOSP-based keyboard and it seems `JNI_LIB_NAME2` is used as a fallback if the normal `JNI_LIB_NAME` doesn't work, or something like that (A.K.A. if `jni_latinimegoogle` isn't installed.) OpenBoard doesn't have that logic in the code, and only reads `JNI_LIB_NAME`

**Comment by stoefelz** [comment]
Is there an (un)official solution to get glide typing working?

**Comment by p4t44** [comment]
> Is there an (un)official solution to get glide typing working?

Not in openboard, in other AOSP-based keyboards, yes

**Comment by JonnyHaystack** [comment]
> > Is there an (un)official solution to get glide typing working?
> 
> Not in openboard, in other AOSP-based keyboards, yes

Are any of them FOSS?

**Comment by p4t44** [comment]
Keyboard itself is, but it requires proprietary libraries for swipe to work.

Only OS keyboard with swipe is ASK.

**Comment by JonnyHaystack** [comment]
Yeah and ASK sadly has been horrible to use in my experience, even without swipe :(

**Comment by JonnyHaystack** [comment]
Does anyone know if Google's proprietary library has any data collection/network accessing code in it?

**Comment by maitra** [comment]
I came across the following zip: [https://forum.xda-developers.com/android/development/fix-cm-aosp-keyboard-swipe-fix-zip-t3182789#post83552617](https://forum.xda-developers.com/android/development/fix-cm-aosp-keyboard-swipe-fix-zip-t3182789#post83552617) but I don't think that it should work anymore, since the binary is for 32-bit and most phones are now arm64, I think. But perhaps it can help as a fix.

**Comment by gochev** [comment]
Shame I would love the gliding even though the only option will be to use a proprietary google library, can we have an experiment with this ? In a fork or something just to see how good or bad it is .. maybe with in-house glide typing as well like in ASK.
I believe this is the best keyboard I really miss the glided typing especially for non-english languages where the words are super long like French or Bulgarian.

**Comment by niks255** [comment]
I don't quite understand why this keyboard doesn't just use swipe typing library that comes with gapps packages. The option is present in AOSP keyboard, the library is distributed separately, so maybe you could sihow the option if the library is present and hide if it isn't? I mean, voice input button doesn't work without a Google's proprietary app as well, at least for me. I don't think that this is that much different.

**Comment by Surendrajat** [comment]
https://github.com/florisboard/florisboard has glide typing now, and it's quite good.

**Comment by maitra** [comment]
> https://github.com/florisboard/florisboard has glide typing now, and it's quite good.

Thanks for this, I have been using this for the last 12 hours. The app itself is on F-Droid so even better. Finally, I got rid of GBoard. Thanks again!

**Comment by niks255** [comment]
> https://github.com/florisboard/florisboard has glide typing now, and it's quite good.

Does it only work with English language currently? I can't get it to work with Russian layout.

**Comment by ildar** [comment]
On Sun, May 23, 2021 at 10:46 PM niks255 wrote:

> Does it only work with English language currently? I can't get it to work
> with Russian layout.
>
this is offtopic HERE.

**Comment by zpangwin** [comment]
> https://github.com/florisboard/florisboard has glide typing now, and it's quite good.

I tried this a few weeks back (whenever it was that F-droid said update available and I saw glide support mentioned in the release notes). Like you say, it wasn't bad. I ultimately ended up going back to gboard though because it wasn't finely tuned or I am fat-fingering things too much for florisboard's glide implementation to handle.

In my case, I suspect the same/similar glide implementation in open-board would work better bc I recall openboard's suggestion support being a fair bit better than florisboard's (at least IMO). To be fair, I only tested this for a week or so right after it came out, so florisboard may have improved on it since then (e.g. if they tweaked sensitivity or improved their suggestion list)

**Comment by Mrnofish** [comment]
> I thought about this, but I'm hesitant to leave this dependency, even on those phones that have Google apps installed. The ideal thing would be to reimplement the swype logic in the library.

It would be ideal, however in the meanwhile having some kind of gesture typing supported would beat hands down not having gesture typing at all.

There is no need to include the library with the app, perhaps gesture typing could have an activation switch defaulting to off and triggering an alert popup, so nobody uses it by mistake.

PS: I'm aware of FlorisBoard however it currently only supports US English, unlike OpenBoard.

**Comment by dslul** [comment]
> It would be ideal, however in the meanwhile having some kind of gesture typing supported would beat hands down not having gesture typing at all.
> 
> There is no need to include the library with the app, perhaps gesture typing could have an activation switch defaulting to off and triggering an alert popup, so nobody uses it by mistake.
> 
> PS: I'm aware of FlorisBoard however it currently only supports US English, unlike OpenBoard.

I can't integrate that library in the app, and even if I could it would kind of defeat the purpose of using this keyboard.

I started experimenting a bit on gesture prediction with openfst. 1-2 years old gboard uses [n-gram FST prediction](https://ai.googleblog.com/2017/05/the-machine-intelligence-behind-gboard.html) ([also here](https://arxiv.org/pdf/1704.03987.pdf)), but training the language models is tricky though, since we obviously can't use people's data. And it will never be on par with current gboard, because now they use [federated learning](https://arxiv.org/pdf/1811.03604.pdf), where they exploit every device computing power to train the model.

**Comment by Mrnofish** [comment]
> I can't integrate that library in the app, and even if I could it would kind of defeat the purpose of using this keyboard.

How come? Early in thread it was said it was a trivial change?

AOSP, which I'm forced to use but has very annoying bugs, can use the library no prob.

It goes without saying that gboard will be better, but perfect ought not be let get in the way of good enough.

**Comment by khimaros** [comment]
FWIW the FlorisBoard  people are discussing opt-in training: https://github.com/florisboard/florisboard/pull/544#issuecomment-847466849

**Comment by HammyHavoc** [comment]
> Hello everyone! I apologise beforehand if this is necro-ing an old issue, but the main reason why OpenBoard doesn't work with Google's swipe library (`libjni_latinimegoogle.so`) was simply due to the fact that the keyboard is expecting the library to have native methods with the package name starting with `org_dslul_openboard`. The library, since it was compiled for AOSP IME (package name starts with `com_android`), won't have that, resulting in a nasty crash (as a result of an `ULE`).
> 
> With that said, I don't at all recommend pursuing this since an effort to bring an open-source solution to gesture typing is preferable.
> 
> For those that are still interested, I have all the relevant commits to work-around the issue at my fork: https://github.com/wordmage/openboard. Note that it's best to include the `libjni_latinimegoogle` file in `jniLibs` as I haven't tested the commits when the libraries are placed under `/system/lib` or other restricted dirs (the main concern here is that Android recently limited the access on dlopen-ing a system library). The changes are very grimy and hacky, so I suggest not deploying them on production hardware.
> 
> Here's a demo of it working: ![Swipe](https://user-images.githubusercontent.com/81974561/136657831-9e0b2eb8-d3b6-454b-93ed-8fd9b59b9486.jpeg)

Is anybody doing any work with this currently?

**Comment by ildar** [comment]
On Sat, Oct 9, 2021 at 6:30 PM @wordmage wrote:

> For those that are still interested, I have all the relevant commits to
> work-around the issue at my fork: https://github.com/wordmage/openboard.

Thanks for this. Yet sharing a (debug) apk would help people try to
understand if it's worth it.
Thanks again.

**Comment by wordmage** [comment]
> Is anybody doing any work with this currently?

I don't expect this to be merged to the mainline OpenBoard repository, seeing that it is a dirty hack and nothing more. I'm open for any suggestions regarding the implementation, however, but I'm not so optimistic it'll ever be in OpenBoard.

> Thanks for this. Yet sharing a (debug) apk would help people try to understand if it's worth it. Thanks again.

I'll be happy to provide this when I have the free time to do so.

**Comment by erkserkserks** [comment]
Hi, for those curious, I built @wordmage's branch and uploaded an apk here: https://github.com/erkserkserks/openboard/releases/tag/v1.4.4-gesture-typing

**Comment by maitra** [comment]
> Hi, for those curious, I built @wordmage's branch and uploaded an apk here: https://github.com/erkserkserks/openboard/releases/tag/v1.4.4-gesture-typing

Works beautifully for me in what I have tried so far. Hope you will have time to put it on F-Droid and maintain it as needed. Thanks!!

**Comment by Mrnofish** [comment]
>  Hope you will have time to put it on F-Droid and maintain it as needed. Thanks!!

+1

While FlorisBoard has just released a new version which the devs are describing as heavily reworked, in the previous version, gesture typing was production level in the sense that it worked reliably, however the accuracy was too lackluster to make it a viable option.

Additionally, language support is limited compared to OpenBoard.

Thus, having a gesture-enabled OpenBoard derivative on F-Droid would be great.

In light of dslul's previous comments, it would seem there's little chance of it materializing otherwise.

**Comment by maitra** [comment]
> > Hope you will have time to put it on F-Droid and maintain it as needed. Thanks!!
> 
> +1
> 
> While FlorisBoard has just released a new version which the devs are describing as heavily reworked, in the previous version, gesture typing was production level in the sense that it worked reliably, however the accuracy was too lackluster to make it a viable option.
> 
> Additionally, language support is limited compared to OpenBoard.
> 
> Thus, having a gesture-enabled OpenBoard derivative on F-Droid would be great.
> 
> In light of dslul's previous comments, it would seem there's little chance of it materializing otherwise.

May have to rename it though, to avoid conflicts and confusion with the original. OpenGlideboard or OpenSwipe or OpenGlide or something more original?

**Comment by LupusArgentum** [comment]
> > Hi, for those curious, I built @wordmage's branch and uploaded an apk here: https://github.com/erkserkserks/openboard/releases/tag/v1.4.4-gesture-typing
> 
> Works beautifully for me in what I have tried so far. Hope you will have time to put it on F-Droid and maintain it as needed. Thanks!!

Yup, please do that, its great to see.

**Comment by Helium314** [comment]
Including proprietary closed source libraries in the app is not going to be accepted on F-Droid.

The only way that *may* work is allowing the user to provide a library via app settings.

**Comment by Surendrajat** [comment]
...or maybe just add it to izzysoft's repo.

**Comment by NoahAndrews** [comment]
Why is a fork necessary? I haven't seen any reason to believe that a PR to add glide typing would be rejected.

**Comment by Mrnofish** [comment]
Glide typing is not necessarily something that I would consider a basic QoL feature, although it could perhaps implemented in a relatively simply fashion - but I'm not going to repeat the whole discussions again.

The reason I used this issue is because some folks had already forked OpenBoard, so it seemed like a decent place to send that message in a bottle.

> I haven't seen any reason to believe that a PR to add glide typing would be rejected.

Then perhaps you should read this issue from the beginning?

The original developer (still part of the openboard-team) stated their wish was to reimplement glide typing from scratch, doing which in a fashion that is considered good for everyday use, is already a huge task on a good day, and obviously won't ever get done if even much smaller things are left to languish for months on end.

Just to mention one such smaller issue that is nevertheless highly irritant, OpenBoard will learn incorrect spellings and there is no way for the user to fix those mistakes without resetting the whole thing.

Another issue I'm encountering on a constant basis is OpenBoard adding spurious extra blanks when making a certain type of correction.

It does seem that there is a PR for this last issue, which however has been left collecting dust for 3 months and is part of roughly 20 other open PRs 

All of this, and additionally the slow release cycle, and sparse activity from openboard-team, suggest that there might be not nearly enough dogfooding of OpenBoard.

So, this is about encouraging those folks with a can-do attitude who already forked and actively worked on OpenBoard, to let them know that a better OpenBoard would be greatly appreciated, if they're willing and able to take up the challenge.

**Comment by Helium314** [comment]
> Personally I could be OK with installing the library manually, but I suspect this would quickly turn into a chore for those running Lineage, as it'd probably entail reinstalling the library every week, whenever an update package is released.

Only saw this now, sorry for the late reply:
No, it would not have to be done again unless you completely reset the app. I'm currently using a version that does excatly this (based on https://github.com/wordmage/openboard with additional preference to load a library).

Such a thing may qualify as acceptable for F-Droid, but allowing to load an arbitrary file to be executed is still a security risk...

> Why is a fork necessary? I haven't seen any reason to believe that a PR to add glide typing would be rejected.

Well, it would considerable re-arrange the locations of code files (see wordmage fork), and it would require a library to be either distributed with the app (closed source, maybe legal issues) or provided by the user (hacky, security issue), or to actually re-implement glide typing (massive amount of work).

> It does seem that there is a PR for this last issue, which however has been left collecting dust for 3 months and is part of roughly 20 other unresolved PRs, among which some being a year old and maybe more.

Most of them are requests for adding dictionaries. I remember those are not merged because they would increase apk size considerably. There also is a PR/branch for allowing to add dictionaries from files that would improve the situation, but it also seems to be collecting dust.

I have the general impression that a) OpenBoard is getting a bunch of new weird bugs/issues with every new Android version and b) the maintainers simply don't have enough time (and maybe new enough Android) to fix those, and even less to add new features.

Actually I have some more changes in my version of OpenBoard, but seeing the slow progress of other PRs I don't want to spend the time necessary to make them ready for a PR (translatable strings, use of preferences instead of hardcoding everything,...)

**Comment by odmfl** [comment]
I notice this commit can help?  It seems that the closed library of google is not used.

https://github.com/LMODroid/platform_packages_inputmethods_LatinIME/commit/bd284873a08bd9dacd8200a255393679e86608f4

Is it good for something or is it a useless commit?

**Comment by odmfl** [comment]
> ![IMG_20220801_121930_714](https://user-images.githubusercontent.com/81974561/182137101-f4aad8d7-ff7b-4ed5-ad23-13bbabf5f609.jpg)
> 
> It works. No Google libraries are added in. I've pushed the necessary changes to [my fork](https://github.com/wordmage/openboard), along with the aforementioned commit. This is much cleaner and I personally think it can be shipped with an OSS build of OpenBoard.

Fantastic, i was convinced it was a good commit.
I wanted to ask you so is gesture typing enabled by default?
In my opinion, however, an option to activate or deactivate it at the user's choice would be useful.
Make a PR, if not, please also add the option to activate / deactivate it and then finally we will be able to close this issue in the best possible way

**Comment by NoahAndrews** [comment]
The new files don't have license blocks, and they include this comment: `// Created by msj on 2017/3/9.` That same comment and matching code can also be found in this repository, which does not even have a license file: https://github.com/shijieKika/GoogleKeyboardV7. Specifically, [this commit](https://github.com/shijieKika/GoogleKeyboardV7/commit/fa1f6d231003237ecc29bb93709a1b1fb7b114ca#diff-d2a1567e679cef7bb58e957c3cad24317370b6899f42761c1495fcef41951619) (fa1f6d231) is from less than a week after the 2017-03-09 date, so it seems fairly likely that that repository was where the code was originally developed.

Since the oldest source of this code that we know of has no license associated with it, we definitely can't assume that it was developed under an open source license.

Please try to be careful about where you get code from.

**Comment by NoahAndrews** [comment]
@nift4 @iKeramat Can you provide details on where the glide typing code came from?

**Comment by nift4** [comment]
You already found the repo - while there is no license file, I simply assumed it is Apache2 licensed. I'm aware that's technically not how it works, and I don't except anyone to merge that code, I personally am happy with this solution and I don't care if the license is not properly defined. The repo has not been taken down in five years, and the commit history very much looks like the chinese developer did the code himself - and he chose to upload it for the public - that's all i cared about when porting those commits.

**Comment by NoahAndrews** [comment]
Also note that the last commit is called `change latinime to kikaime`, and that the owner of that GitHub account has an `@kikatech.com` email address. http://www.kikatech.com/ is the website for a closed-source still-maintained Android keyboard. This feels much more like an employee being careless about where they put company code than it does someone developing open source code and forgetting to add a license file.

**Comment by nift4** [comment]
It's probably nothing to merge as is from the POV of an big project like OpenBoard indeed. Maybe someone will reimplement it based on the ideas of this example code, or maybe a fork with the patch will have to do.
I personally don't care, it works, that's all that matters to me.

**Comment by nift4** [comment]
On a slightly related note, most new files are partially modified copies of the typing/ folder variants, not all-new code.

**Comment by odmfl** [comment]
Yes but the published code is open source on github, it has no license and therefore I don't see where the problem lies and in the end it works. This kikatek keyboard is closed source, but the code we are interested in is opensource and there is no license or anything else that forbids to use it

> Also note that the last commit is called `change latinime to kikaime`, and that the owner of that GitHub account has an `@kikatech.com` email address. http://www.kikatech.com/ is the website for a closed-source still-maintained Android keyboard. This feels much more like an employee being careless about where they put company code than it does someone developing open source code and forgetting to add a license file.

**Comment by NoahAndrews** [comment]
At least by US law (not sure about other jurisdictions) code without an explicit license has all rights reserved, which means you can't use it.

**Comment by Mrnofish** [comment]
@Helium314 thanks for the answer, very interesting.

> I remember those are not merged because they would increase apk size considerably. There also is a PR/branch for allowing to add dictionaries from files that would improve the situation, but it also seems to be collecting dust.

I could name at least another FOSS keyboard that has solved that problem by distributing the languages as optional components.

Sad to see OpenBoard is still languishing, though, it's still my daily driver.

**Comment by odmfl** [comment]
Ok, I didn't know this US law, even if we are not US.
However the solution is one, let's see if the user is still active.
@shijieKika @ShijieKika if you are active for pleasure can you give us explanations about it?  Or enter a license in your reposity?

> At least by US law (not sure about other jurisdictions) code without an explicit license has all rights reserved, which means you can't use it. Open source is not the default.

**Comment by BBO-repo** [comment]
> Hi, for those curious, I built @wordmage's branch and uploaded an apk here: https://github.com/erkserkserks/openboard/releases/tag/v1.4.4-gesture-typing

Works like a charm!
Used on pixel 6A with [GrapheneOS](https://grapheneos.org/), in both French and English languages
Thank so much to all contributors !!!

**Comment by rugk** [comment]
@wordmage 's @erkserkserks Could the gesture fork be ported to v1.4.5 of OpenBoard, respectively could the chnages be merged in there?

https://github.com/openboard-team/openboard/releases/tag/v1.4.5
